axes/handlers/cache.py

```diff
@@ -0,0 +1,116 @@
+from logging import getLogger
+
+from axes.conf import settings
+from axes.exceptions import AxesSignalPermissionDenied
+from axes.handlers.base import AxesBaseHandler
+from axes.signals import user_locked_out
+from axes.utils import (
+    get_axes_cache,
+    get_client_cache_key,
+    get_client_ip_address,
+    get_client_path_info,
+    get_client_str,
+    get_client_username,
+    get_client_user_agent,
+    get_credentials,
+    get_cool_off,
+)
+
+log = getLogger(settings.AXES_LOGGER)
+
+
+class AxesCacheHandler(AxesBaseHandler):  # pylint: disable=too-many-locals
+    """
+    Signal handler implementation that records user login attempts to cache and locks users out if necessary.
+    """
+
+    def get_failures(self, request, credentials=None, attempt_time=None) -> int:
+        cache = get_axes_cache()
+        cache_key = get_client_cache_key(request, credentials)
+        return cache.get(cache_key, default=0)
+
+    def user_login_failed(self, sender, credentials, request=None, **kwargs):  # pylint: disable=too-many-locals
+        """
+        When user login fails, save attempt record in cache and lock user out if necessary.
+
+        :raises AxesSignalPermissionDenied: if user should be locked out.
+        """
+
+        if request is None:
+            log.error('AXES: AxesCacheHandler.user_login_failed does not function without a request.')
+            return
+
+        username = get_client_username(request, credentials)
+        ip_address = get_client_ip_address(request)
+        user_agent = get_client_user_agent(request)
+        path_info = get_client_path_info(request)
+        client_str = get_client_str(username, ip_address, user_agent, path_info)
+
```

Comment:
I wonder if these could be placed into a single helper function

Reply:
They could and most probably should. One good place to load and set these would be the middleware, but I was considering if that is the best place since some people could opt to use their own middleware or skip the Axes one. 

Fetching these where they are used imposes the least state, and we would ideally like this handler chain to be as stateless as possible, so a single helper function is probably the best place.